feat(codegen): xamarin codegen

LopezArnaud opened the pull request:
Fully supported :
 - Labels
 - Border
 - Background
 - Rectangle and round shapes
 - Code indentation

shikanime committed:
chore: enrich parser test (#284)

* test: refactor parser name

* chore: remove missed type todo

* test: add parse point test

* test: add parse set text test

* test: add set solid test

* test: reorder sketch parser test

* chore: update sketchapp type

t-dambacher committed:
Merge pull request #1 from xlayers/master

chore: enrich parser test (#284)

manekinekko committed:
feat: make the 3d design rotate (#288)

Fixes #128

krvajal committed:
test(e2e): add cypress test for landing page (#285)

* test(e2e): add cypress test for landing page

* chore(cypress): add scripts to run cypresss tests

* chore(tests): prevent Jest from running the e2e tests

* ci(e2e): add step to run e2e test on CircleCI

* chore(cypress): more cypress to dependencies

* chore: fix test:e2e run script

* chore: add cypress artifacts to the gitignore file

* chore: update caching strategy on the test job

* ci: update base image for the test job

* test: update home spec

* refactor: rename npm-scripts

* fix: add missing comma

* fix: rename e2e job on CI

* fix: rename cy.open

shikanime committed:
chore: add svg mask support (#271)

* chore: add group mask

* chore: add shape test

* fix: types

* chore: enforce SketchMSData type cast

* fix: fixed size parse point

manekinekko committed:
docs: update readme with intro section

shikanime committed:
fix: remove svg from shape property (#291)

shikanime committed:
refactor: renaming files and element tag (#289)

* refactor: renaming files and element tag

* refactor: rename sketch classes

* chore: delete the misplaced file

* refactor: namespacing viewer component

* refactor: remove sketch prefex on upload

* chore: fix viewer canvas tag

* fix: viewer page tag

* chore: remove residual files

shikanime committed:
refactor: move home outside app scope (#290)

manekinekko committed:
fix: update 3d tooltip label

manekinekko committed:
refactor: move back to a standard angular app (#292)

LopezArnaud committed:
feat(codegen): xamarin codegen

Fully supported :
 - Labels
 - Border
 - Background
 - Rectangle and round shapes
 - Code indentation

t-dambacher committed:
Merge pull request #2 from xlayers/master

Merge latest changes

LopezArnaud committed:
feat(codegen): xamarin codegen, shape recognition

now recognise rectangles, lines and circles svg and replace them with corresponding xaml element

t-dambacher commented on packages/xlayers/src/app/editor/code/editor-container/codegen/codegen.service.ts:
```diff
@@ -81,15 +81,30 @@ export class CodeGenService {
       } else if (file.language.includes('css')) {
         comment.start = '/*';
         comment.end = '*/';
+      } else if (file.language.includes('xaml')) {
```
As the content of this `if` statement is the same as the one a few lines above, you can combine both conditions and have only one block, like this :
```
      if (file.language.includes('html') || file.language.includes('xaml')) {
        comment.start = '<!--';
        comment.end = '-->';
      } else if (file.language.includes('css')) {
        comment.start = '/*';
        comment.end = '*/';
      }
```

t-dambacher commented on packages/xlayers/src/app/editor/code/editor-container/codegen/xamarin-forms/codegen/codegenvisitor.service.ts:
```diff
@@ -11,37 +11,103 @@ export abstract class CodeGenVisitor {
+    protected abstract openTag(tag: string): string;
+    protected abstract closeTag(tag: string): string;
+
+
```
You here have an empty line that you can remove.

t-dambacher commented on packages/xlayers/src/app/editor/code/editor-container/codegen/xamarin-forms/codegen/codegenvisitor.service.ts:
```diff
@@ -11,37 +11,103 @@ export abstract class CodeGenVisitor {
+        + Math.round(green * 255).toString(16).padStart(2, '0')
+        + Math.round(blue * 255).toString(16).padStart(2, '0');
+    }
+    // return colorhex: string or false
```
You can add a new line, to have a more visual separation between both methods, like in the rest of the codebase.

t-dambacher commented on packages/xlayers/src/app/editor/code/editor-container/codegen/xamarin-forms/codegen/shape.service.ts:
```diff
@@ -0,0 +1,197 @@
+        return Math.round(number * 100 + 0.001) / 100;
+    }
+
+    distance(point: Point): number {
```
What are the differences between all those methods ? Their naming could be improved, if they all used. 
Also, this is nitpicking, but you should add an empty line between all methods.

t-dambacher commented on packages/xlayers/src/app/editor/code/editor-container/codegen/xamarin-forms/codegen/shape.service.ts:
```diff
@@ -0,0 +1,197 @@
+// class to recognise if sketchapp shape is a rectangle, a line, a circle(TODO) or something else
```
It seems this file has a lot of snake casing. I may be mistaken, but I didn't have the impression this was a common practice in the project, and that camel casing was privileged. Could you check please ?

t-dambacher commented on packages/xlayers/src/app/editor/code/editor-container/codegen/xamarin-forms/codegen/shape.service.ts:
```diff
@@ -0,0 +1,197 @@
+    x: number;
+    y: number;
+
+    constructor(str_or_x, y = null) {
```
Shouldn't we specify the types excepted, in all methods ?

LopezArnaud replied:
You are right i am going to review the types declarations

t-dambacher commented on packages/xlayers/src/app/editor/code/editor-container/codegen/xamarin-forms/codegen/xamarin-forms-codegenvisitor.service.ts:
```diff
@@ -1,49 +1,141 @@
+    const attributes = [];
+
+    let content;
+    // bait the transcrcipter
```
Nitpicking: there's a typo
```suggestion
    // bait the transcripter
```

t-dambacher commented on packages/xlayers/src/app/editor/code/editor-container/codegen/xamarin-forms/codegen/xamarin-forms-codegenvisitor.service.ts:
```diff
@@ -1,49 +1,141 @@
-        template.push(this.indent(depth + 1, this.closeTag('Grid.ColumnDefinitions')));
+  protected visitText(ast: SketchMSLayer): string {
+    return `<Label Text="${ast.attributedString.string}"
+       FontSize="${ast.attributedString.attributes[0].attributes.MSAttributedStringFontAttribute.attributes.size}"
```
This whole method could be simplified a bit, by using intermediate variables.

t-dambacher commented on packages/xlayers/src/app/editor/code/editor-container/codegen/xamarin-forms/codegen/xamarin-forms-codegenvisitor.service.ts:
```diff
@@ -1,49 +1,141 @@
+    const a = new Shape((ast as any).points);
 
-        template.push(this.indent(depth, this.closeTag('Grid')));
+    if (a.is_round()) {
```
This whole method could be more readable by splitting it by shape. `visitShape` could just perform the general logic, switch between the shape types, and redirect the rendering logic to other functions, each specialized in a specific shape (`visitRectangleShape`, `visitLineShape`, etc.)

Jefiozie committed:
feat: new framework selection (#297)

LopezArnaud committed:
feat(codegen): xamarin codegen Svg

correction on circle recognition,
add some test on shape.service
refactoring
implements the use of svg in xamarin.forms
minor changes on parser
  - svg fill color rgba to hex
  - update coordonates notation to fit xaml requirements

LopezArnaud committed:
test(parser): update parser test

update parser tests to match changes done in last commit

t-dambacher committed:
Merge pull request #4 from xlayers/master

feat: new framework selection (#297)

LopezArnaud committed:
fix(codegen): close Label tags

LopezArnaud committed:
Merge pull request #2 from t-dambacher/master

Merge master

LopezArnaud committed:
Merge branch 'master-update-clean' into add-xamarin-forms-support-merge

Conflicts:
	packages/xlayers/src/app/editor/code/editor-container/editor-container.component.html
	tests/integration/__snapshots__/sketch-parsing.spec.ts.snap

LopezArnaud committed:
refactor(codegen): merge xamarinForms support

LopezArnaud committed:
test(parser): update parser test

update parser tests to match changes done to svg for xamarinForms support
